### DISCUSSION TITLE
Is there still interest in add Code Lens?

### DISCUSSION THREAD

**Discussion by DevelopmentCool2449** [joaotavora/eglot#1376] [Q&A]
Hello @joaotavora.

Is there still any interest in adding code lens to eglot?
To be specific, about this PR https://github.com/joaotavora/eglot/pull/71

If there is still interest in adding Code Lens but not from this PR https://github.com/joaotavora/eglot/pull/71, I would like to help implement this in my own patch (If the assignment is necessary, I would gladly do it).

**Comment by joaotavora** [comment]
Sure, in general there is interest. But how do you plan to design it exactly?  If you don't want to explain, just submit your patch to bug-gnu-emacs@gnu.org, but I think it would benefit from an explanation of the strategy (overlays, buffers, UX, etc)

**Reply by DevelopmentCool2449** [reply]
> Sure, in general there is interest. But how do you plan to design it exactly?

Similar in how https://github.com/joaotavora/eglot/pull/71 is done.
I'll see if i can use inlay hint as base for this.

>  I think it would benefit from an explanation of the strategy (overlays, buffers, UX, etc)

Yeah, in fact this discussions can be used for explain or suggest how to implement this.

**Reply by DevelopmentCool2449** [reply]
Ok i think i know how i can make this,
my ideas are using eglot inlay hints code as base for implement code lens
this include the overlays and async, 
also i'm planing add support for mouse clicks,

also i'm planning add functions similar to https://github.com/joaotavora/eglot/pull/71 as eglot-next-lens and eglot-previous-lens
about the keybindings i dont know to which keys must i bind.

When i can make this implementation stable i'll move this discussion to a PR for make it easy review.

**Reply by joaotavora** [reply]
OK, that's still a bit vague, but go ahead.   A mockup would go a long way in explaining at least the UI/UX -- which is really the most important thing here -- but maybe it's harder to do than implement an actual prototype.

**Reply by DevelopmentCool2449** [reply]
> OK, that's still a bit vague, but go ahead. A mockup would go a long way in explaining at least the UI/UX -- which is really the most important thing here -- but maybe it's harder to do than implement an actual prototype.

You are right, i've taken the time to make a few mockups about my idea.

In this mock up code lens acts like inlay hints, however with codelens you can click them, for example,
in `refs` you must be able to click them and find all the references with xref.
![Captura2](https://github.com/joaotavora/eglot/assets/163072436/24cdb3b1-fc89-4892-9411-2d9cd4dd2df4)

![Captura](https://github.com/joaotavora/eglot/assets/163072436/4e469480-7f7c-4ef8-a8c8-ccf2b85a9063)
(i couldn't get the mouse cursor in this mockup but i think the idea gets the point)

Also in eglot menu add something like `eglot-code-lens-actions` for executing lens actions if avalaible
(i couldn't make for this one a mockup sorry)

Also, where i can find the first inlay hints implementation commit? i've read eglot.el code for understanding inlay hints but i can't get how does it work.
Any help is appreciated.

**Comment by ypytomets** [comment]
AFAIK, code lenses is the current way how test cases supposed to be discovered by lsp: https://github.com/emacs-lsp/lsp-metals/issues/75
FYI there's some related outdated snippet: https://gist.github.com/Gavinok/fa23bbea7f44725eb633c002a8aa803a
***
BTW, is it possible to use:
- [discoverTest](https://github.com/microsoft/language-server-protocol/issues/1267) (e.g. [test](https://scalameta.org/metals/docs/editors/vscode/#via-test-explorer) [explorer](https://scalameta.org/metals/docs/integrations/test-explorer/))
- [runnables](https://github.com/microsoft/language-server-protocol/issues/944) ?
